tests/devices/qutrit_mixed/test_qutrit_mixed_apply_operation.py

```diff
@@ -0,0 +1,426 @@
+
+import os
+import pytest
+import numpy as np
+from scipy.stats import unitary_group
+import pennylane as qml
+from pennylane import math
+from pennylane.operation import Channel
+from pennylane.devices.qutrit_mixed.apply_operation import (
+    apply_operation_einsum,
+    apply_operation,
+)
+
+ml_frameworks_list = [
+    "numpy",
+    pytest.param("autograd", marks=pytest.mark.autograd),
+    pytest.param("jax", marks=pytest.mark.jax),
+    pytest.param("torch", marks=pytest.mark.torch),
+    pytest.param("tensorflow", marks=pytest.mark.tf),
+]
+
+methods = [apply_operation_einsum, apply_operation]  # TODO add tensordot back when added
+broadcasting_methods = [apply_operation_einsum, apply_operation]
+subspaces = [(0, 1), (0, 2), (1, 2)]
+
+
+def load_data(num_qutrits, batched):
+    file_name_start = f"{num_qutrits}_qutrit"
+    if num_qutrits != "one":
+        file_name_start += "s"
+
+    file_name = f"{file_name_start}_states.npy"
+    path = os.path.join(os.getcwd(), "test_ref_files", file_name)
+    data = np.load(path, allow_pickle=True)
+    if batched:
+        return data
+    return data[0]
+
+
+one_qutrit_state = pytest.fixture(lambda: load_data("one", False), scope="module")
+two_qutrits_state = pytest.fixture(lambda: load_data("two", False), scope="module")
+two_qutrits_states = pytest.fixture(lambda: load_data("two", True), scope="module")
+three_qutrits_state = pytest.fixture(lambda: load_data("three", False), scope="module")
+three_qutrits_states = pytest.fixture(lambda: load_data("three", True), scope="module")
+
+
+def test_custom_operator_with_matrix(one_qutrit_state):
+    """Test that apply_operation works with any operation that defines a matrix."""
+    mat = np.array(
+        [
+            [-0.35546532 - 0.03636115j, -0.19051888 - 0.38049108j, 0.07943913 - 0.8276115j],
+            [-0.2766807 - 0.71617593j, -0.1227771 + 0.61271557j, -0.0872488 - 0.11150285j],
+            [-0.2312502 - 0.47894201j, -0.04564929 - 0.65295532j, -0.3629075 + 0.3962342j],
+        ]
+    )
+
+    # pylint: disable=too-few-public-methods
+    class CustomOp(qml.operation.Operation):
+        num_wires = 1
+
+        def matrix(self):
+            return mat
+
+    new_state = apply_operation(CustomOp(0), one_qutrit_state)
+    assert qml.math.allclose(new_state, mat @ one_qutrit_state @ np.conj(mat).T)
+
+
+@pytest.mark.parametrize("ml_framework", ml_frameworks_list)
+@pytest.mark.parametrize("method", methods)
+@pytest.mark.parametrize("wire", (0, 1))
+class TestTwoQubitStateSpecialCases:
+    """Test the special cases on a two qutrit state.  Also tests the special cases for einsum application method
+    for additional testing of these generic matrix application methods."""
+
+    # Currently not added as special cases, but will be in future
+    # TODO add tensordot back after it has been added
+
+    def test_TAdd(self, method, wire, ml_framework, two_qutrits_state):
```

Comment:
These tests seem overly complicated and can make future changes difficult. The way I see it, there are two things that these tests need to validate:

* Check that the operation is allowed correctly. This is pretty straightforward, just need to validate similarly to other device tests that the density matrix after applying an operation is as expected. Using a state will all non-zero amplitudes is a good idea for this.
* Check that the custom implementation is used (once added). This can be done by spying on `apply_operation_einsum` with `mocker.spy` and validating that it doesn't get called. You can check out how mocker is used in other test files.

Reply:
Oh good idea! I guess these can probably be removed then.

Reply:
@mudit2812 I removed the tests for now and left a stub for a TODO